moto/kms/models.py

```diff
@@ -714,5 +731,84 @@ def get_public_key(self, key_id: str) -> Tuple[Key, bytes]:
         key = self.describe_key(key_id)
         return key, key.private_key.public_key()
 
+    def rotate_key_on_demand(self, key_id: str) -> str:
+        key: Key = self.keys[self.get_key_id(key_id)]
+
+        rotation = {
+            "KeyId": key_id,
+            "RotationDate": datetime.now().timestamp(),
+            "RotationType": "ON_DEMAND",
+        }
+
+        # Add to key rotations
+        key.rotations.append(rotation)
+
+        return key_id
+
+    def list_key_rotations(
+        self, key_id: str, limit: int, marker: Union[str, None]
+    ) -> Tuple[List[Dict[str, Union[str, float]]], bool, Union[str, None]]:
+        key: Key = self.keys[self.get_key_id(key_id)]
+
+        # we treat the marker as an index-offset
+        marker = marker or "0"
+        index_offset = int(marker)
+
+        rotations_after_offset = key.rotations[index_offset:]
+
+        if len(rotations_after_offset) > limit:
+            next_marker = str(index_offset + limit)
+            rotations = rotations_after_offset[:limit]
+            truncated = True
+        else:
+            next_marker = None
+            rotations = rotations_after_offset
+            truncated = False
+
+        return rotations, truncated, next_marker
+
+    def generate_mac(
+        self,
+        message: bytes,
+        key_id: str,
+        mac_algorithm: str,
+        grant_tokens: List[str],
+        dry_run: bool,
+    ) -> Tuple[str, str, str]:
+        key = self.keys[key_id]
+
+        if (
+            key.key_usage != "GENERATE_VERIFY_MAC"
+            or key.key_spec not in KeySpec.hmac_key_specs()
+        ):
+            raise InvalidKeyUsageException()
+
+        mac = generate_hmac(
+            key=key.key_material, message=message, mac_algorithm=mac_algorithm
+        )
+        return mac, mac_algorithm, key_id
+
+    def verify_mac(
+        self,
+        message: bytes,
+        key_id: str,
+        mac_algorithm: str,
+        mac: str,
+        grant_tokens: List[str],
+        dry_run: bool,
+    ) -> Tuple[str, bool, str]:
+        regenerated_mac, _, _ = self.generate_mac(
+            message=message,
+            key_id=key_id,
+            mac_algorithm=mac_algorithm,
+            grant_tokens=grant_tokens,
+            dry_run=dry_run,
+        )
+
+        if mac != regenerated_mac:
+            raise KMSInvalidMacException()
+
+        return key_id, True, mac_algorithm
```

Comment:
Do we need to return anything here? We already know the `key_id` and `mac_algorithm`, and the `is_valid` will always be `True` - so we do not return any new information. Not returning anything would simplify this method

Reply:
Makes sense, I shouldn't have blindly followed the scaffold-layout :)

I'll update it.